Input Handling: ignore KeyAlt state on character input handling

gucio321 opened the pull request:
#1 claims, that doing ctrl==alt will enable AltGr behaviour on some keyboard layouts.
As long as this might have been true for some particular cases this wasn't for mine.

As of my testing, when AltGr is pressed, only the io.KeyAlt is set to true and KeyCtrl remains false.
The perfect solution would be to use
`ImGui::IsKeyPressedi(ImGuiKeyAltLeft) && !ctrl`
However I'm not sure about what exactly might be interpreted as AltGr on some keyboard layouts. Since characters input handling itself comes from different part of imgui (so a special charracter will be posted to the queue when the valid ALtGr key was pressed in the first place), I think that skipping `!alt` check in this if statement should work just fine.

Should fix #28, ref https://github.com/AllenDang/cimgui-go/issues/470

gucio321 committed:
Input Handling: ignore KeyAlt state on character input handling

 #1 claims, that doing ctrl==alt will enable AltGr behaviour on some keyboard layouts.
As long as this might have been true for some particular cases this wasn't for mine.

As of my testing, when AltGr is pressed, only the io.KeyAlt is set to true
and KeyCtrl remains false.
The perfect solution would be to use
`ImGui::IsKeyPressedi(ImGuiKeyAltLeft) && !ctrl`
However I'm not sure about what exactly might be interpreted as AltGr on
some keyboard layouts. Since characters input handling itself comes from
different part of imgui (so a special charracter will be posted to the
queue when the valid ALtGr key was pressed in the first place), I think
that skipping `!alt` check in this if statement should work just fine.

Should fix #28, ref https://github.com/AllenDang/cimgui-go/issues/470

santaclose commented:
so we are going back to what we had before?

gucio321 commented:
I just looked closer at #1 and: yes,. Idk what keyboard layout does @aeris170 have, but this change seems to break the editor for anyone else... Maybe we can check both conditions and `||` them...

aeris170 commented:
Hey @gucio321 , what exactly did #1 break for you? You shouldn't be pressing either CTRL or ALT during character typing anyways (unless you use German/Turkish layout). Could you clarify why you need the merge undone? I don't understand what's wrong with it.

gucio321 commented:
In a polish layout, when we want to type eg. `ąęś` etc we press alt+a etc. This doesn't work now. Also I heard voices that german kb layout doesn't work either (see linked issue in this rapper and the one that I have linked in cimgui-go)

aeris170 commented:
I think I understand the topic now. Here's what I took from it. 

![Notes_250928_082107_1.jpg](https://github.com/user-attachments/assets/76981dfa-de4e-4c10-9124-efc9d5994c5b)

![Notes_250928_082107_2.jpg](https://github.com/user-attachments/assets/c1da8866-3fd9-46c2-878e-cf7bb842482e)


Neither the original `!ctrl`  nor the current `ctrl == alt` is correct. I can test with a Turkish layout, are you okay with testing this expression on your end @santaclose @gucio321 ?

X is don't care.

aeris170 commented:
Can't comment on why German wouldn't work. Also, ImGui's `AltGr` handling might have changed since we merged the PR so we could be going back to what it was, too. When I opened that PR, `AltGr` was triggering both `Ctrl` and `Alt` in ImGui. There is so much to test here. Will report back in about 15 hours. Ping me if I forget.

aeris170 commented:
Well, now that I think of it, it can't be ImGui's handling. Because in a Turkish keyboard you can't use `Alt` to type `[`, you can only use `AltGr`. Is that the same in Polish keyboard, or can you use both `Alt` and `AltGr` @gucio321 ?

gucio321 commented:
Its AltGr too on polish kb

> Because in a Turkish keyboard you can't use Alt to type [, you can only use AltGr

Now I think it works the same way on every layout (?)

Now the question is why the version from this pr doesn't work for you @aeris170 ...
What's so special about turkish kb what makes ctrl state involved?

aeris170 commented:
@gucio321 `Ctrl` is used to distinguish between `Alt` and `AltGr`. There has to be some distinction. Otherwise, `Alt` would also type characters `AltGr` types.

gucio321 commented:
Well if that's what your test says I am not going to argue it just doesn't work this way on my end. 
For me it's like when you press Ctrl + leftAlt + anything it does not behave like an altgr. If you press however rightAlt+anything it inserts special characters. 
However the thing we are referring to as an `alt` in this PR is actually `io.KeyAlt` and for me this becomes true for both: left and right alt while the `io.KeyCtrl` remains false. 
I think that characters input handling happens somewhere deeper in the imgui and it doesn't involve iO structure. (Well this was just for documentation purposes 😁)

Now: the condition you proposed will also work for me so I think I will just replace it in this PR is it okay for you?

gucio321 commented:
@aeris170 could you check if the condition of form `!ctrl && !ImGui::IsKeyDown(ImGuiKey_LeftAlt)` works for you?

aeris170 commented:
@gucio321 Sorry for the delayed response.

I have now tested and I see that ImGui indeed changed how it processes the `AltGr` key in `ImGuiIO` at some point after #1 has been merged. The `AltGr` key in Turkish layout used to set both `ImGui::GetIO().KeyCtrl` and `ImGui::GetIO().KeyAlt` to true, which is indeed true on a keyboard layout level. I'm both able to use `AltGr` and `Ctrl+Alt` to type `[` anywhere (which is `AltGr+8` for the interested).

This is no longer the case with ImGui 1.91.9 (probably earlier too). The `AltGr` key being `Ctrl+Alt` in Turkish layout is still true but it is no longer how `ImGuiIO` treats it. This change in `ImGui` is what probably broke everything and lead us to this point. The funny thing is, since `AltGr` no longer triggers `ImGui::GetIO().KeyCtrl`, I can't use `AltGr+8` to type `[` anymore to the `TextEditor`! 🤣🤣🤣

The old old way of `!ctrl` wouldn't work with Turkish layout. It would block `Ctrl+Alt+8` combination, which is rare but is still equivalent to `AltGr+8`. The expression `!(ctrl && !alt)` works for both `Ctrl+Alt` and `AltGr`, could you test it for your layouts? @gucio321 @santaclose

aeris170 commented:
The expression `!ctrl && !ImGui::IsKeyDown(ImGuiKey_LeftAlt)` (because it contains `!ctrl`) preserves `AltGr` functionality but breaks `Ctrl+Alt` (which is `AltGr` in Turkish layout).

gucio321 committed:
input handling: update the condition to handle AltGr correctly on the Turkish kb layout

Thank you @aeris170 for helping me while figuring out the right
condition that will work on a Turkish keyboard layout.

gucio321 commented:
ok, it works for me!
@santaclose I think this is a final version of this PR. COuld you test and review please ;-)

santaclose commented:
works fine, should we squash the commits?

aeris170 commented:
> works fine, should we squash the commits?

Yeah, squash and merge!

santaclose commented:
huh, didn't know it was an option from the web interface 👍🏼

krux02 commented:
I am sorry, I can not follow this discussion. No combination of alt+ctrl will produce an AltGr modifier, AltGr is a separate modifier, that works independent of ctrl,alt,super,shift

aeris170 commented:
@krux02 on which layout?

krux02 commented:
> @krux02 on which layout?

German (`setxkbme de` if you are on Linux). I wouldn't be able to use emacs at all, if any modifier compinations would produce an `alt gr` key instead.

gucio321 commented:
But as far as I know you the imguiio structure doesn't distinguish between Alt and altgr.

krux02 commented:
I think all imgui needs to support is a char send event from the os. AltGr is afaik no modifier at all, if the only options are alt,str,super,shift,meta,hyper. It's just not a modifier supposed to be tracked by the application logic. And if it's tracked.
